ton_client/src/abi/tests.rs

```diff
@@ -343,3 +353,184 @@ fn decode_v2() {
     };
     assert_eq!(expected, decode_events("te6ccgEBAQEAVQAApeACvg5/pmQpY4m61HmJ0ne+zjHJu3MNG8rJxUDLbHKBu/AAAAAAAAAMKr6z6rxK3xYJAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABA"));
 }
+
+#[test]
+fn test_is_empty_pubkey() -> Result<()> {
+    let pubkey = ed25519_dalek::PublicKey::from_bytes(&[0; 32])?;
+
+    assert!(is_empty_pubkey(&pubkey));
+
+    let pubkey = ed25519_dalek::PublicKey::from_bytes(&[1; 32])?;
+    assert!(!is_empty_pubkey(&pubkey));
+
+    let mut array = [0; 32];
+    array[0] = 1;
+    let pubkey = ed25519_dalek::PublicKey::from_bytes(&array)?;
+    assert!(!is_empty_pubkey(&pubkey));
+
+    Ok(())
+}
+
+#[test]
+fn test_resolve_pubkey() -> Result<()> {
+    let tvc = base64::encode(include_bytes!("../tests/contracts/abi_v2/Hello.tvc"));
+    let mut deploy_set = DeploySet {
+        tvc: tvc.clone(),
+        ..Default::default()
+    };
+    let mut image = create_tvc_image("", None, &tvc)?;
+    assert!(resolve_pubkey(&deploy_set, &image, &None )?.is_none());
+
+    let external_pub_key = Some("1234567890ABCDEF1234567890ABCDEF1234567890ABCDEF1234567890ABCDEF".to_owned());
+    let resolved = resolve_pubkey(&deploy_set, &image, &external_pub_key)?;
+    assert_eq!(resolved, external_pub_key);
+
+    let resolved = resolve_pubkey(&deploy_set, &image, &external_pub_key)?;
+
+    assert_eq!(resolved, external_pub_key);
+
+    let tvc_pubkey_empty = ed25519_dalek::PublicKey::from_bytes(&[0; 32])?;
+    image.set_public_key(&tvc_pubkey_empty)?;
+
+    let resolved = resolve_pubkey(&deploy_set, &image, &external_pub_key)?;
+
+    assert_eq!(resolved, external_pub_key);
+
+    let tvc_pubkey_1 = ed25519_dalek::PublicKey::from_bytes(&[1; 32])?;
+    image.set_public_key(&tvc_pubkey_1)?;
+
+    let resolved = resolve_pubkey(&deploy_set, &image, &external_pub_key)?;
+
+    assert_eq!(resolved, Some(hex::encode(tvc_pubkey_1.as_bytes())));
+
+    let initial_pub_key = Some("1234567890123456789012345678901234567890123456789012345678901234".to_owned());
+    deploy_set.initial_pubkey = initial_pub_key.clone();
+
+    let resolved = resolve_pubkey(&deploy_set, &image, &external_pub_key)?;
+
+    assert_eq!(resolved, initial_pub_key);
+
+    Ok(())
+}
+
+#[tokio::test]
+async fn test_encode_message_pubkey() -> Result<()> {
+    let context = Arc::new(ClientContext::new(Default::default())?);
```

Comment:
Do not call functions directly and don't use ClientContext. Call testing functions via TestClient. See existing tests to understand how it should be implemented.